File: src/main/java/at/hrechny/predictionsbot/service/predictor/CompetitionService.java
package at.hrechny.predictionsbot.service.predictor;

import at.hrechny.predictionsbot.connector.apifootball.ApiFootballConnector;
import at.hrechny.predictionsbot.connector.apifootball.exception.ApiFootballConnectorException;
import at.hrechny.predictionsbot.connector.apifootball.model.Fixture;
import at.hrechny.predictionsbot.connector.apifootball.model.Status;
import at.hrechny.predictionsbot.connector.apifootball.model.Team;
import at.hrechny.predictionsbot.database.entity.CompetitionEntity;
import at.hrechny.predictionsbot.database.entity.MatchEntity;
import at.hrechny.predictionsbot.database.entity.RoundEntity;
import at.hrechny.predictionsbot.database.entity.SeasonEntity;
import at.hrechny.predictionsbot.database.entity.TeamEntity;
import at.hrechny.predictionsbot.database.model.MatchStatus;
import at.hrechny.predictionsbot.database.model.RoundType;
import at.hrechny.predictionsbot.database.repository.CompetitionRepository;
import at.hrechny.predictionsbot.database.repository.MatchRepository;
import at.hrechny.predictionsbot.database.repository.SeasonRepository;
import at.hrechny.predictionsbot.database.repository.TeamRepository;
import at.hrechny.predictionsbot.exception.FixturesSynchronizationException;
import at.hrechny.predictionsbot.exception.NotFoundException;
import at.hrechny.predictionsbot.exception.RequestValidationException;
import at.hrechny.predictionsbot.mapper.CompetitionMapper;
import at.hrechny.predictionsbot.mapper.SeasonMapper;
import at.hrechny.predictionsbot.model.Competition;
import at.hrechny.predictionsbot.model.Season;
import jakarta.transaction.Transactional;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.regex.Pattern;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@Transactional
@RequiredArgsConstructor
public class CompetitionService {

  private final SeasonMapper seasonMapper;
  private final SeasonRepository seasonRepository;
  private final CompetitionMapper competitionMapper;
  private final CompetitionRepository competitionRepository;
  private final TeamRepository teamRepository;
  private final MatchRepository matchRepository;
  private final ApiFootballConnector apiFootballConnector;

  public UUID addCompetition(Competition competition) {
    log.info("Adding the new competition: {}", competition);
    var entity = competitionRepository.save(competitionMapper.modelToEntity(competition));
    log.info("The competition has been successfully stored: {}", entity.getId());
    return entity.getId();
  }

  public Competition getCompetition(UUID competitionId) {
    CompetitionEntity entity = competitionRepository.findById(competitionId)
        .orElseThrow(() -> new NotFoundException("Competition with the ID " + competitionId + " not found"));
    return competitionMapper.entityToModel(entity);
  }

  public List<Competition> getCompetitions() {
    var entityList = competitionRepository.findAll();
    var competitionList = new ArrayList<Competition>();
    entityList.forEach(entity -> competitionList.add(competitionMapper.entityToModel(entity)));
    return competitionList;
  }

  public UUID addSeason(UUID competitionId, Season season) {
    log.info("Adding the new season for the competition {}: {}", competitionId, season);
    CompetitionEntity competitionEntity = competitionRepository.findById(competitionId)
        .orElseThrow(() -> new NotFoundException("Competition with the ID " + competitionId + " not found"));

    validateActiveSeasons(competitionId, season);

    var seasonEntity = seasonRepository.save(seasonMapper.modelToEntity(competitionEntity, season));
    log.info("The season has been successfully stored");

    return seasonEntity.getId();
  }

  public void updateSeason(UUID competitionId, Season season) {
    log.info("Updating the season {} for the competition {}", season.getId(), competitionId);
    var seasonEntity = seasonRepository.findById(season.getId()).orElse(null);
    if (seasonEntity == null || !seasonEntity.getCompetition().getId().equals(competitionId)) {
      throw new NotFoundException("Season with the ID " + season.getId() + " not found");
    }

    validateActiveSeasons(competitionId, season);

    seasonMapper.updateEntity(seasonEntity, season);
    seasonRepository.save(seasonEntity);
    log.info("The season {} has been successfully updated", seasonEntity.getId());
  }

  public SeasonEntity getSeason(UUID seasonId) {
    return seasonRepository.findById(seasonId).orElseThrow(() -> new NotFoundException("Season " + seasonId + " not found"));
  }

  public List<Season> getSeasons(UUID competitionId) {
    var entityList = seasonRepository.findAllByCompetitionId(competitionId);
    var seasonList = new ArrayList<Season>();
    entityList.forEach(entity -> seasonList.add(seasonMapper.entityToModel(entity)));
    return seasonList;
  }

  public SeasonEntity getCurrentSeason(UUID competitionId) {
    return seasonRepository.findFirstByCompetitionIdAndActiveIsTrue(competitionId)
        .orElseThrow(() -> new NotFoundException("No active season found for the competition " + competitionId));
  }

  public RoundEntity getUpcomingRound(UUID competitionId) {
    return matchRepository.findUpcoming(getCurrentSeason(competitionId)).map(MatchEntity::getRound).orElse(null);
  }

  public List<MatchEntity> getFixtures(Instant from, Instant to) {
    return matchRepository.findAllByStartTimeAfterAndStartTimeBeforeOrderByStartTimeAsc(from, to);
  }

  public RoundEntity getRound(UUID competitionId, Integer orderNumber) {
    var season = getCurrentSeason(competitionId);
    return season.getRounds().stream().filter(roundEntity -> orderNumber.equals(roundEntity.getOrderNumber())).findFirst().orElse(null);
  }

  public List<SeasonEntity> getActiveSeasons() {
    log.info("Starting to refresh fixtures data for the all active competitions");
    return seasonRepository.findAllByActiveIsTrue();
  }

  public void refreshActiveFixtures(UUID seasonId) {
    var seasonEntity = getSeason(seasonId);
    var activeMatches = matchRepository.findAllActive(seasonEntity);
    if (activeMatches.isEmpty()) {
      return;
    }

    var fixturesIds = new ArrayList<Long>();
    activeMatches.forEach(match -> fixturesIds.add(match.getApiFootballId()));
    try {
      var fixtures = apiFootballConnector.getFixtures(fixturesIds);
      refreshFixtures(fixtures, seasonEntity);
    } catch (ApiFootballConnectorException ex) {
      log.error("Failed to refresh fixtures: {}", ex.getMessage());
    }
  }

  public void refreshFixtures(SeasonEntity seasonEntity) {
    log.info("Start refreshing fixtures data for the season {}", seasonEntity.getId());
    try {
      var fixtures = apiFootballConnector.getFixtures(seasonEntity.getCompetition().getApiFootballId(), seasonEntity.getYear());
      refreshFixtures(fixtures, seasonEntity);
    } catch (ApiFootballConnectorException ex) {
      log.error("Failed to refresh fixtures for {}: {}", seasonEntity.getCompetition().getName(), ex.getMessage());
    }
  }

  private void refreshFixtures(List<Fixture> fixtures, SeasonEntity seasonEntity) {
    var rounds = seasonEntity.getRounds();
    var matches = rounds.stream().flatMap(round -> round.getMatches().stream()).toList();

    fixtures.forEach(fixture -> {
      var fixtureData = fixture.getFixture();
      var score = fixture.getScore().getFulltime().getHome() != null ? fixture.getScore().getFulltime() : fixture.getGoals();
      var roundList = getRound(rounds, fixture.getLeague().getRound());
      if (roundList.isEmpty()) {
        refreshRounds(seasonEntity);
        roundList = getRound(rounds, fixture.getLeague().getRound());
      }

      MatchEntity matchEntity;
      var existingMatchEntity = matches.stream().filter(match -> match.getApiFootballId().equals(fixtureData.getId())).findFirst();
      if (existingMatchEntity.isPresent()) {
        matchEntity = existingMatchEntity.get();
      } else {
        matchEntity = new MatchEntity();
        matchEntity.setApiFootballId(fixtureData.getId());
        matchEntity.setHomeTeam(getTeamEntity(fixture.getTeams().getHome()));
        matchEntity.setAwayTeam(getTeamEntity(fixture.getTeams().getAway()));
      }

      // update round if needed
      var round = getRound(roundList, matchEntity.getHomeTeam(), matchEntity.getAwayTeam());
      if (matchEntity.getRound() == null) {
        matchEntity.setRound(round);
        round.getMatches().add(matchEntity);
      } else if (!matchEntity.getRound().equals(round)) {
        matchEntity.getRound().getMatches().remove(matchEntity);
        matchEntity.setRound(round);
        round.getMatches().add(matchEntity);
      }

      // update match status and results
      matchEntity.setHomeTeamScore(score.getHome());
      matchEntity.setAwayTeamScore(score.getAway());
      matchEntity.setStatus(mapStatus(fixture.getFixture().getStatus()));
      if (fixtureData.getDate() != null && matchEntity.getStatus() != MatchStatus.NOT_DEFINED) {
        matchEntity.setStartTime(fixtureData.getDate().toInstant());
      } else {
        matchEntity.setStartTime(null);
      }
    });
    seasonRepository.save(seasonEntity);
    log.info("Fixtures have been successfully updated for the season {}", seasonEntity.getId());
  }

  private MatchStatus mapStatus(Status status) {
    if (status == null || status.getStatus() == null) {
      return MatchStatus.NOT_DEFINED;
    }

    return switch (status.getStatus()) {
      case NS -> MatchStatus.PLANNED;
      case _1H, HT, _2H, LIVE -> MatchStatus.STARTED;
      case ABD, CANC, INT, PST, SUSP, WO, TBD -> MatchStatus.NOT_DEFINED;
      case AET, P, PEN, ET, AWD, BT, FT -> MatchStatus.FINISHED;
    };
  }

  private TeamEntity getTeamEntity(Team team) {
    var teamEntityOptional = teamRepository.findFirstByApiFootballId(team.getId());
    if (teamEntityOptional.isPresent()) {
      var teamEntity = teamEntityOptional.get();
      // if no changes for the team detected, return the existing entity, otherwise update it
      if (Objects.equals(teamEntity.getName(), team.getName()) && Objects.equals(teamEntity.getLogoUrl(), team.getLogo())) {
        return teamEntity;
      }
      teamEntity.setName(team.getName());
      teamEntity.setLogoUrl(team.getLogo());
      teamEntity = teamRepository.save(teamEntity);
      log.info("Team {} has been updated: {}", teamEntity.getName(), teamEntity.getId());
      return teamEntity;
    } else {
        return createTeam(team);
    }
  }

  private TeamEntity createTeam(Team team) {
    var teamEntity = new TeamEntity();
    teamEntity.setName(team.getName());
    teamEntity.setApiFootballId(team.getId());
    teamEntity.setLogoUrl(team.getLogo());
    teamEntity = teamRepository.save(teamEntity);
    log.info("New team {} has been created: {}", teamEntity.getName(), teamEntity.getId());
    return teamEntity;
  }

  private void validateActiveSeasons(UUID competitionId, Season season) {
    if (season.isActive()) {
      int activeSeasons = seasonRepository.countAllByActiveIsTrueAndCompetitionId(competitionId);
      if (activeSeasons > 0) {
        throw new RequestValidationException("Not possible to add more than one active season");
      }
    }
  }

  private void refreshRounds(SeasonEntity seasonEntity) {
    var actualRounds = apiFootballConnector.getRounds(seasonEntity.getCompetition().getApiFootballId(), seasonEntity.getYear());
    var roundEntities = seasonEntity.getRounds();
    var lastRound = roundEntities.stream().max(Comparator.comparingInt(RoundEntity::getOrderNumber)).orElse(null);
    AtomicInteger nextOrderNumber = new AtomicInteger(lastRound != null ? lastRound.getOrderNumber() + 1 : 1);
    for (var round : actualRounds) {
      if (roundEntities.stream().noneMatch(roundEntity -> roundEntity.getApiFootballId().equals(round))) {
        RoundType.getByAlias(round).forEach(roundType -> {
          var roundEntity = new RoundEntity();
          roundEntity.setType(roundType);
          roundEntity.setOrderNumber(getOrderNumber(round, roundType, nextOrderNumber));
          roundEntity.setApiFootballId(round);
          roundEntity.setSeason(seasonEntity);
          roundEntities.add(roundEntity);
        });
      }
    }
  }

  private int getOrderNumber(String roundName, RoundType roundType, AtomicInteger nextOrderNumber) {
    var pattern = Pattern.compile("^(.+) - (\\d+)$").matcher(roundName);
    if (pattern.matches()) {
      return Integer.parseInt(pattern.group(2));
    }
    return roundType == RoundType.QUALIFYING ? 0 : nextOrderNumber.getAndIncrement();
  }

  private List<RoundEntity> getRound(List<RoundEntity> rounds, String apiFootballId) {
    return rounds.stream().filter(roundEntity -> roundEntity.getApiFootballId().equals(apiFootballId)).toList();
  }

  private RoundEntity getRound(List<RoundEntity> roundList, TeamEntity homeTeam, TeamEntity awayTeam) {
    if (roundList.isEmpty()) {
      throw new FixturesSynchronizationException("Round could not be find for the match between " + homeTeam.getName() + " and " + awayTeam.getName());
    }

    if (roundList.size() == 1) {
      return roundList.get(0);
    }

    if (roundList.size() == 2) {
      RoundEntity firstRound;
      RoundEntity returnRound;

      if (roundList.get(0).getType().name().endsWith("RETURN")) {
        firstRound = roundList.get(1);
        returnRound = roundList.get(0);
      } else {
        firstRound = roundList.get(0);
        returnRound = roundList.get(1);
      }

      var returnMatch = firstRound.getMatches().stream()
          .anyMatch(matchEntity -> matchEntity.getHomeTeam().equals(awayTeam) && matchEntity.getAwayTeam().equals(homeTeam));
      if (returnMatch) {
        return returnRound;
      } else {
        return firstRound;
      }
    }

    throw new FixturesSynchronizationException("Unexpected number of rounds found for the match between "
        + homeTeam.getName() + " and " + awayTeam.getName() + ": " + roundList);
  }
}

File: src/main/kotlin/at/hrechny/predictionsbot/service/predictor/CompetitionService.kt
package at.hrechny.predictionsbot.service.predictor

import at.hrechny.predictionsbot.connector.ApiConnector
import at.hrechny.predictionsbot.connector.model.FixtureSyncDto
import at.hrechny.predictionsbot.connector.model.FixtureSyncStatus
import at.hrechny.predictionsbot.connector.model.RoundSyncDto
import at.hrechny.predictionsbot.connector.model.RoundSyncType
import at.hrechny.predictionsbot.connector.model.TeamSyncDto
import at.hrechny.predictionsbot.controller.model.connector.ApiConnectorIdRequestDto
import at.hrechny.predictionsbot.controller.model.connector.ApiConnectorIdResponseDto
import at.hrechny.predictionsbot.controller.model.competition.CompetitionCreateRequestDto
import at.hrechny.predictionsbot.controller.model.competition.CompetitionResponseDto
import at.hrechny.predictionsbot.controller.model.competition.SeasonCreateRequestDto
import at.hrechny.predictionsbot.controller.model.competition.SeasonResponseDto
import at.hrechny.predictionsbot.controller.model.competition.SeasonUpdateRequestDto
import at.hrechny.predictionsbot.database.entity.ApiConnectorIdEntity
import at.hrechny.predictionsbot.database.entity.CompetitionEntity
import at.hrechny.predictionsbot.database.entity.MatchEntity
import at.hrechny.predictionsbot.database.entity.RoundEntity
import at.hrechny.predictionsbot.database.entity.SeasonEntity
import at.hrechny.predictionsbot.database.entity.TeamEntity
import at.hrechny.predictionsbot.database.model.ApiConnectorEntityType
import at.hrechny.predictionsbot.database.model.MatchStatus
import at.hrechny.predictionsbot.database.model.RoundType
import at.hrechny.predictionsbot.database.repository.CompetitionRepository
import at.hrechny.predictionsbot.database.repository.MatchRepository
import at.hrechny.predictionsbot.database.repository.SeasonRepository
import at.hrechny.predictionsbot.database.repository.TeamRepository
import at.hrechny.predictionsbot.exception.FixturesSynchronizationException
import at.hrechny.predictionsbot.exception.NotFoundException
import at.hrechny.predictionsbot.exception.RequestValidationException
import at.hrechny.predictionsbot.mapper.CompetitionMapper
import at.hrechny.predictionsbot.mapper.SeasonMapper
import jakarta.inject.Singleton
import io.micronaut.transaction.annotation.Transactional
import java.time.Instant
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger
import java.util.regex.Pattern
import org.slf4j.LoggerFactory

@Singleton
@Transactional
open class CompetitionService(
    private val seasonMapper: SeasonMapper?,
    private val seasonRepository: SeasonRepository?,
    private val competitionMapper: CompetitionMapper?,
    private val competitionRepository: CompetitionRepository?,
    private val teamRepository: TeamRepository?,
    private val matchRepository: MatchRepository?,
    private val apiConnector: ApiConnector?,
    private val apiConnectorService: ApiConnectorService?,
) {
    open fun addCompetition(competition: CompetitionCreateRequestDto): UUID {
        log.info("Adding the new competition: {}", competition)
        val entity = competitionRepository!!.save(competitionMapper!!.modelToEntity(competition))
        log.info("The competition has been successfully stored: {}", entity.id)
        return entity.id!!
    }

    open fun getCompetition(competitionId: UUID): CompetitionResponseDto {
        val entity: CompetitionEntity = competitionRepository!!.findById(competitionId)
            .orElseThrow { NotFoundException("Competition with the ID $competitionId not found") }
        return toCompetitionResponse(entity)
    }

    open fun getCompetitions(): List<CompetitionResponseDto> =
        competitionRepository!!.findAll().map(::toCompetitionResponse)

    open fun addConnectorId(
        connectorCode: String,
        entityType: ApiConnectorEntityType,
        internalId: UUID,
        request: ApiConnectorIdRequestDto,
    ): ApiConnectorIdResponseDto {
        requireInternalEntity(entityType, internalId)
        return connectorService().upsertId(
            connectorCode,
            entityType,
            request.connectorEntityId!!,
            internalId,
        ).toApiConnectorIdResponse()
    }

    open fun getConnectorIds(
        entityType: ApiConnectorEntityType,
        internalId: UUID,
    ): List<ApiConnectorIdResponseDto> =
        connectorService()
            .findConnectorIdMappings(internalId, entityType)
            .map { entity -> entity.toApiConnectorIdResponse() }

    open fun addSeason(competitionId: UUID, season: SeasonCreateRequestDto): UUID {
        log.info("Adding the new season for the competition {}: {}", competitionId, season)
        val competitionEntity = competitionRepository!!.findById(competitionId)
            .orElseThrow { NotFoundException("Competition with the ID $competitionId not found") }

        validateActiveSeasons(competitionId, season)
        val seasonEntity = seasonRepository!!.save(seasonMapper!!.modelToEntity(competitionEntity, season))
        log.info("The season has been successfully stored")
        return seasonEntity.id!!
    }

    open fun updateSeason(competitionId: UUID, season: SeasonUpdateRequestDto) {
        log.info("Updating the season {} for the competition {}", season.id, competitionId)
        val seasonEntity = seasonRepository!!.findById(season.id!!).orElse(null)
        if (seasonEntity == null || seasonEntity.competition!!.id != competitionId) {
            throw NotFoundException("Season with the ID ${season.id} not found")
        }

        validateActiveSeasons(competitionId, season)
        seasonMapper!!.updateEntity(seasonEntity, season)
        seasonRepository.save(seasonEntity)
        log.info("The season {} has been successfully updated", seasonEntity.id)
    }

    open fun getSeason(seasonId: UUID): SeasonEntity =
        seasonRepository!!.findById(seasonId).orElseThrow { NotFoundException("Season $seasonId not found") }

    open fun getSeasons(competitionId: UUID): List<SeasonResponseDto> =
        seasonRepository!!.findAllByCompetitionId(competitionId).map(seasonMapper!!::entityToModel)

    open fun getCurrentSeason(competitionId: UUID): SeasonEntity =
        seasonRepository!!.findFirstByCompetitionIdAndActiveIsTrue(competitionId)
            .orElseThrow { NotFoundException("No active season found for the competition $competitionId") }

    open fun getUpcomingRound(competitionId: UUID): RoundEntity? =
        matchRepository!!.findUpcoming(getCurrentSeason(competitionId)).map(MatchEntity::round).orElse(null)

    open fun getFixtures(from: Instant, to: Instant): List<MatchEntity> =
        matchRepository!!.findAllByStartTimeAfterAndStartTimeBeforeOrderByStartTimeAsc(from, to)

    open fun getRound(competitionId: UUID, orderNumber: Int): RoundEntity? {
        val season = getCurrentSeason(competitionId)
        return season.rounds.firstOrNull { roundEntity -> orderNumber == roundEntity.orderNumber }
    }

    open fun getActiveSeasons(): List<SeasonEntity> {
        log.info("Starting to refresh fixtures data for the all active competitions")
        return seasonRepository!!.findAllByActiveIsTrue()
    }

    open fun refreshActiveFixtures(seasonId: UUID) {
        val seasonEntity = getSeason(seasonId)
        val activeMatches = matchRepository!!.findAllActive(seasonEntity)
        if (activeMatches.isEmpty()) {
            return
        }

        val connectorName = connectorName()
        val fixtureIds = activeMatches.map { match ->
            connectorService().requireConnectorEntityId(
                connectorName,
                ApiConnectorEntityType.MATCH,
                match.id!!,
            )
        }
        val fixtures = apiConnector!!.getFixturesByExternalIds(fixtureIds)
        refreshFixtures(fixtures, seasonEntity, FixtureSyncPolicy.UPDATE_ONLY)
    }

    open fun refreshFixtures(seasonEntity: SeasonEntity) {
        val managedSeasonEntity = getSeason(seasonEntity.id!!)
        log.info("Start refreshing fixtures data for the season {}", managedSeasonEntity.id)
        val competitionExternalId = connectorService().requireConnectorEntityId(
            connectorName(),
            ApiConnectorEntityType.COMPETITION,
            managedSeasonEntity.competition!!.id!!,
        )
        val fixtures = apiConnector!!.getSeasonFixtures(
            competitionExternalId,
            managedSeasonEntity.year!!,
        )
        refreshFixtures(fixtures, managedSeasonEntity, FixtureSyncPolicy.AUTHORITATIVE_BOOTSTRAP)
    }

    private fun refreshFixtures(
        fixtures: List<FixtureSyncDto>,
        seasonEntity: SeasonEntity,
        syncPolicy: FixtureSyncPolicy,
    ) {
        val rounds = requireKnownMappings(fixtures, seasonEntity, syncPolicy)
        val matches = rounds.flatMap { round -> distinctMatchEntities(round.matches) }
            .distinctBy(::matchEntityKey)
        val matchConnectorIds = mutableMapOf<MatchEntity, String>()

        fixtures.forEach { fixture ->
            val roundList = getRound(rounds, fixture.round)

            val matchEntity = findMatchByConnectorId(matches, seasonEntity, fixture.externalId)
                ?: run {
                    if (syncPolicy == FixtureSyncPolicy.UPDATE_ONLY) {
                        throw FixturesSynchronizationException(
                            "No internal match found for connector ${connectorName()} fixture ${fixture.externalId}",
                        )
                    }
                    val homeTeam = getTeamEntity(fixture.homeTeam)
                    val awayTeam = getTeamEntity(fixture.awayTeam)
                    findMatchByCanonicalIdentity(matches, roundList, homeTeam, awayTeam)
                        ?: MatchEntity().apply {
                            this.homeTeam = homeTeam
                            this.awayTeam = awayTeam
                        }
                }

            val round = getRound(roundList, matchEntity)
            if (matchEntity.round == null) {
                matchEntity.round = round
                round.matches.add(matchEntity)
            } else if (matchEntity.round != round) {
                matchEntity.round!!.matches.remove(matchEntity)
                matchEntity.round = round
                round.matches.add(matchEntity)
            }

            matchEntity.homeTeamScore = fixture.score.home
            matchEntity.awayTeamScore = fixture.score.away
            matchEntity.status = mapStatus(fixture.status)
            matchEntity.startTime = if (fixture.startTime != null && matchEntity.status != MatchStatus.NOT_DEFINED) {
                fixture.startTime
            } else {
                null
            }
            matchConnectorIds[matchEntity] = fixture.externalId
        }
        seasonRepository!!.save(seasonEntity)
        recordMatchConnectorIds(matchConnectorIds)
        log.info("Fixtures have been successfully updated for the season {}", seasonEntity.id)
    }

    private fun mapStatus(status: FixtureSyncStatus): MatchStatus =
        when (status) {
            FixtureSyncStatus.PLANNED -> MatchStatus.PLANNED
            FixtureSyncStatus.STARTED -> MatchStatus.STARTED
            FixtureSyncStatus.NOT_DEFINED -> MatchStatus.NOT_DEFINED
            FixtureSyncStatus.FINISHED -> MatchStatus.FINISHED
        }

    private fun requireKnownMappings(
        fixtures: List<FixtureSyncDto>,
        seasonEntity: SeasonEntity,
        syncPolicy: FixtureSyncPolicy,
    ): List<RoundEntity> {
        var rounds = distinctRoundEntities(seasonEntity.rounds)
        val missingRoundLabels = fixtures
            .map(FixtureSyncDto::round)
            .filter { round -> getRound(rounds, round).isEmpty() }
            .map(RoundSyncDto::name)
            .distinct()
        if (missingRoundLabels.isNotEmpty() && syncPolicy == FixtureSyncPolicy.AUTHORITATIVE_BOOTSTRAP) {
            refreshRounds(seasonEntity)
            rounds = distinctRoundEntities(seasonEntity.rounds)
        }

        val missingMappings = MissingConnectorMappings()
        fixtures
            .map(FixtureSyncDto::round)
            .filter { round -> getRound(rounds, round).isEmpty() }
            .map(RoundSyncDto::name)
            .distinct()
            .forEach(missingMappings::addRound)

        fixtures
            .asSequence()
            .map(FixtureSyncDto::homeTeam)
            .plus(fixtures.asSequence().map(FixtureSyncDto::awayTeam))
            .distinctBy(TeamSyncDto::externalId)
            .filter { team ->
                connectorService()
                    .findInternalId(connectorName(), ApiConnectorEntityType.TEAM, team.externalId)
                    .isEmpty
            }
            .forEach(missingMappings::addTeam)

        if (syncPolicy == FixtureSyncPolicy.UPDATE_ONLY) {
            fixtures
                .map(FixtureSyncDto::externalId)
                .distinct()
                .filter { externalId ->
                    connectorService()
                        .findInternalId(connectorName(), ApiConnectorEntityType.MATCH, externalId)
                        .isEmpty
                }
                .forEach(missingMappings::addMatch)
        }

        missingMappings.throwIfNotEmpty(connectorName(), seasonEntity)
        return rounds
    }

    private fun getTeamEntity(team: TeamSyncDto): TeamEntity {
        val internalId = connectorService()
            .findInternalId(connectorName(), ApiConnectorEntityType.TEAM, team.externalId)
            .orElse(null)
        if (internalId != null) {
            var teamEntity = teamRepository!!.findById(internalId)
                .orElseThrow { NotFoundException("Team mapping $internalId points to a missing team") }
            if (teamEntity.name == team.name && teamEntity.logoUrl == team.logoUrl) {
                return teamEntity
            }
            teamEntity.name = team.name
            teamEntity.logoUrl = team.logoUrl
            teamEntity = teamRepository.save(teamEntity)
            log.info("Team {} has been updated: {}", teamEntity.name, teamEntity.id)
            return teamEntity
        }
        throw FixturesSynchronizationException(
            "No internal team found for connector ${connectorName()} team ${team.externalId}",
        )
    }

    private fun validateActiveSeasons(competitionId: UUID, season: SeasonCreateRequestDto) {
        if (season.isActive()) {
            val activeSeasons = seasonRepository!!.countAllByActiveIsTrueAndCompetitionId(competitionId)
            if (activeSeasons > 0) {
                throw RequestValidationException("Not possible to add more than one active season")
            }
        }
    }

    private fun refreshRounds(seasonEntity: SeasonEntity) {
        val competitionExternalId = connectorService().requireConnectorEntityId(
            connectorName(),
            ApiConnectorEntityType.COMPETITION,
            seasonEntity.competition!!.id!!,
        )
        val actualRounds = apiConnector!!.getRounds(
            competitionExternalId,
            seasonEntity.year!!,
        )
        val roundEntities = seasonEntity.rounds
        val lastRound = roundEntities.maxByOrNull(RoundEntity::orderNumber)
        val nextOrderNumber = AtomicInteger(if (lastRound != null) lastRound.orderNumber + 1 else 1)
        for (round in actualRounds) {
            round.types.map(::mapRoundType).forEach { roundType ->
                val orderNumber = round.orderNumber ?: getOrderNumber(round.name, roundType, nextOrderNumber)
                if (roundEntities.none { roundEntity -> sameRound(roundEntity, roundType, round.orderNumber, orderNumber) }) {
                    roundEntities.add(
                        RoundEntity().apply {
                            type = roundType
                            this.orderNumber = orderNumber
                            this.season = seasonEntity
                        },
                    )
                }
            }
        }
    }

    private fun validateActiveSeasons(competitionId: UUID, season: SeasonUpdateRequestDto) {
        if (season.isActive()) {
            val activeSeasons = seasonRepository!!.countAllByActiveIsTrueAndCompetitionId(competitionId)
            if (activeSeasons > 0) {
                throw RequestValidationException("Not possible to add more than one active season")
            }
        }
    }

    private fun getOrderNumber(roundName: String, roundType: RoundType, nextOrderNumber: AtomicInteger): Int {
        val pattern = Pattern.compile("^(.+) - (\\d+)$").matcher(roundName)
        if (pattern.matches()) {
            return pattern.group(2).toInt()
        }
        return if (roundType == RoundType.QUALIFYING) 0 else nextOrderNumber.getAndIncrement()
    }

    private fun getRound(rounds: List<RoundEntity>, round: RoundSyncDto): List<RoundEntity> {
        val roundTypes = round.types.map(::mapRoundType).toSet()
        return rounds.filter { roundEntity ->
            roundEntity.type in roundTypes && (round.orderNumber == null || roundEntity.orderNumber == round.orderNumber)
        }.distinctBy(::roundEntityKey)
    }

    private fun sameRound(
        roundEntity: RoundEntity,
        roundType: RoundType,
        sourceOrderNumber: Int?,
        resolvedOrderNumber: Int,
    ): Boolean =
        roundEntity.type == roundType && (sourceOrderNumber == null || roundEntity.orderNumber == resolvedOrderNumber)

    private fun mapRoundType(roundType: RoundSyncType): RoundType =
        RoundType.valueOf(roundType.name)

    private fun findMatchByConnectorId(
        matches: List<MatchEntity>,
        seasonEntity: SeasonEntity,
        externalId: String,
    ): MatchEntity? {
        val internalId = connectorService()
            .findInternalId(connectorName(), ApiConnectorEntityType.MATCH, externalId)
            .orElse(null)
            ?: return null
        val match = matches.firstOrNull { existingMatch -> existingMatch.id == internalId }
            ?: matchRepository!!.findById(internalId)
                .orElseThrow { NotFoundException("Match mapping $internalId points to a missing match") }
        if (match.round?.season?.id != seasonEntity.id) {
            throw FixturesSynchronizationException(
                "Connector match $externalId points to match ${match.id} outside season ${seasonEntity.id}",
            )
        }
        return match
    }

    private fun findMatchByCanonicalIdentity(
        matches: List<MatchEntity>,
        roundList: List<RoundEntity>,
        homeTeam: TeamEntity,
        awayTeam: TeamEntity,
    ): MatchEntity? {
        val matched = matches.filter { match ->
            match.round in roundList && match.homeTeam == homeTeam && match.awayTeam == awayTeam
        }
        if (matched.size > 1) {
            throw FixturesSynchronizationException(
                "Multiple matches found for ${homeTeam.name} vs ${awayTeam.name} in connector round $roundList",
            )
        }
        return matched.firstOrNull()
    }

    private fun recordMatchConnectorIds(matchConnectorIds: Map<MatchEntity, String>) {
        val connectorName = connectorName()
        matchConnectorIds.forEach { (match, externalId) ->
            if (match.id != null) {
                connectorService().upsertId(connectorName, ApiConnectorEntityType.MATCH, externalId, match.id!!)
            }
        }
    }

    private fun toCompetitionResponse(entity: CompetitionEntity): CompetitionResponseDto =
        competitionMapper!!.entityToModel(entity)

    private fun ApiConnectorIdEntity.toApiConnectorIdResponse(): ApiConnectorIdResponseDto =
        ApiConnectorIdResponseDto().apply {
            connectorCode = this@toApiConnectorIdResponse.connectorCode
            entityType = this@toApiConnectorIdResponse.entityType
            connectorEntityId = this@toApiConnectorIdResponse.connectorEntityId
            internalId = this@toApiConnectorIdResponse.internalId
        }

    private fun connectorService(): ApiConnectorService =
        apiConnectorService ?: throw IllegalStateException("ApiConnectorService is required")

    private fun connectorName(): String = apiConnector!!.name

    private fun requireInternalEntity(entityType: ApiConnectorEntityType, internalId: UUID) {
        val exists = when (entityType) {
            ApiConnectorEntityType.COMPETITION -> competitionRepository!!.findById(internalId).isPresent
            ApiConnectorEntityType.SEASON -> seasonRepository!!.findById(internalId).isPresent
            ApiConnectorEntityType.MATCH -> matchRepository!!.findById(internalId).isPresent
            ApiConnectorEntityType.TEAM -> teamRepository!!.findById(internalId).isPresent
        }
        if (!exists) {
            throw NotFoundException("Internal $entityType $internalId not found")
        }
    }

    private fun getRound(roundList: List<RoundEntity>, match: MatchEntity): RoundEntity {
        val existingRound = match.round
        if (existingRound != null && roundList.any { round -> sameRoundEntity(round, existingRound) }) {
            return existingRound
        }
        return getRound(roundList, match.homeTeam!!, match.awayTeam!!)
    }

    private fun getRound(roundList: List<RoundEntity>, homeTeam: TeamEntity, awayTeam: TeamEntity): RoundEntity {
        if (roundList.isEmpty()) {
            throw FixturesSynchronizationException("Round could not be find for the match between ${homeTeam.name} and ${awayTeam.name}")
        }

        if (roundList.size == 1) {
            return roundList[0]
        }

        if (roundList.size == 2) {
            val firstRound: RoundEntity
            val returnRound: RoundEntity
            if (roundList[0].type!!.name.endsWith("RETURN")) {
                firstRound = roundList[1]
                returnRound = roundList[0]
            } else {
                firstRound = roundList[0]
                returnRound = roundList[1]
            }

            val returnMatch = firstRound.matches.any { matchEntity ->
                matchEntity.homeTeam == awayTeam && matchEntity.awayTeam == homeTeam
            }
            return if (returnMatch) returnRound else firstRound
        }

        throw FixturesSynchronizationException(
            "Unexpected number of rounds found for the match between ${homeTeam.name} and ${awayTeam.name}: $roundList",
        )
    }

    private fun distinctRoundEntities(rounds: Collection<RoundEntity>): List<RoundEntity> =
        rounds.distinctBy(::roundEntityKey)

    private fun distinctMatchEntities(matches: Collection<MatchEntity>): List<MatchEntity> =
        matches.distinctBy(::matchEntityKey)

    private fun sameRoundEntity(left: RoundEntity, right: RoundEntity): Boolean =
        if (left.id != null && right.id != null) {
            left.id == right.id
        } else {
            left === right
        }

    private fun roundEntityKey(round: RoundEntity): Any =
        round.id ?: System.identityHashCode(round)

    private fun matchEntityKey(match: MatchEntity): Any =
        match.id ?: System.identityHashCode(match)

    private class MissingConnectorMappings {
        private val rounds = linkedSetOf<String>()
        private val teams = linkedSetOf<String>()
        private val matches = linkedSetOf<String>()

        fun addRound(roundName: String) {
            rounds.add(roundName)
        }

        fun addTeam(team: TeamSyncDto) {
            teams.add("${team.name ?: "unknown"} (${team.externalId})")
        }

        fun addMatch(externalId: String) {
            matches.add(externalId)
        }

        fun throwIfNotEmpty(connectorName: String, seasonEntity: SeasonEntity) {
            if (rounds.isEmpty() && teams.isEmpty() && matches.isEmpty()) {
                return
            }

            throw FixturesSynchronizationException(
                buildString {
                    append("Missing connector mappings for ")
                    append(connectorName)
                    append(" season ")
                    append(seasonEntity.id)
                    append(": ")
                    appendSection("rounds", rounds)
                    appendSection("teams", teams)
                    appendSection("matches", matches)
                },
            )
        }

        private fun StringBuilder.appendSection(name: String, values: Set<String>) {
            if (values.isEmpty()) {
                return
            }
            if (!endsWith(": ")) {
                append("; ")
            }
            append(name)
            append("=")
            append(values.joinToString(", "))
        }
    }

    private companion object {
        val log = LoggerFactory.getLogger(CompetitionService::class.java)
    }
}
